### PR TITLE
upsample2d function rough draft

### PR DESCRIPTION
**Pull request by naruto-raj**
Hi Team,
i am pretty new to contributing in opensource projects. Please have a review of the upsample2d function and let me know of anything that is required or should be changed. the function is added in objax.function.ops module.

### CODE REVIEW

File: objax/functional/core/ops.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 __all__ = ['dynamic_slice', 'pad', 'rsqrt', 'stop_gradient', 'top_k',
-           'flatten', 'one_hot', 'upscale_nn']
+           'flatten', 'one_hot', 'upscale_nn', 'upsample_2d']
```
Comment (david-berthelot) [inline objax/functional/core/ops.py:16]: Alphabetical order please.

File: objax/functional/core/ops.py [hunk 2]
```diff
@@ -55,3 +55,22 @@ def upscale_nn(x: JaxArray, scale: int = 2) -> JaxArray:
     x = x.reshape(s[:2] + (s[2], 1, s[3], 1))
     x = jn.tile(x, (1, 1, 1, scale, 1, scale))
     return x.reshape(s[:2] + (scale * s[2], scale * s[3]))
+
+
+def upsample_2d(x: JaxArray, scale: tuple=(2,2), method: str = 'bilinear') -> JaxArray:
```
Comment (david-berthelot) [inline objax/functional/core/ops.py]: Order APIs alphabetically too please.

Comment (david-berthelot) [reply objax/functional/core/ops.py]: For `scale`, you could use IntOrTuple to make it easier for users.
See `k` for example in Conv2D https://github.com/google/objax/blob/master/objax/nn/layers.py#L144

Comment (david-berthelot) [reply objax/functional/core/ops.py]: For method use an EnumOrString, see `padding` argument in Conv2D for inspiration: https://github.com/google/objax/blob/master/objax/nn/layers.py#L148

File: objax/functional/core/ops.py [hunk 3]
```diff
@@ -55,3 +55,22 @@ def upscale_nn(x: JaxArray, scale: int = 2) -> JaxArray:
...
+            Args:
+                x: input tensor.
+                scale: tuple which contains the scaling factor
+                method: either of the two interpolation methods ['bilinear', 'nearest'].
+            returns:
+                upscaled 2d image tensor
+            """
+    if method not in {'nearest', 'bilinear'}:
+        raise ValueError('`method` argument should be one of `"nearest"` '
+                         'or `"bilinear"`.')
```
Comment (david-berthelot) [inline objax/functional/core/ops.py:71]: With an enum, this will work without having to write a check.

**Review by david-berthelot — CHANGES_REQUESTED** [review]
We recently a guide for new contributors:
https://objax.readthedocs.io/en/latest/dev/adding_module.html

Check the parts on documentation and tests in particular, but you can read it all too of course, it's not too long.

Thanks!

File: objax/util/util.py [hunk 4]
```diff
@@ -164,3 +165,13 @@ def to_tuple(v: Union[Tuple[Number, ...], Number, Iterable], n: int):
         return (v,) * n
     else:
         return tuple(v)
+
+
+def to_upsample(upsample: Union[UpSample, str]) \
```
Comment (david-berthelot) [inline objax/util/util.py:170]: Can you put an `assert` at the end to inform the user they didn't pass an Upsample or str argument?

File: objax/constants.py [hunk 5]
```diff
@@ -21,3 +21,9 @@ class ConvPadding(enum.Enum):
     """An Enum holding the possible padding values for convolution modules."""
     SAME = 'SAME'
     VALID = 'VALID'
+
+
+class UpSample(enum.Enum):
```
Comment (david-berthelot) [inline objax/constants.py:26]: There are more methods available: https://jax.readthedocs.io/en/latest/_autosummary/jax.image.resize.html

Could you add them?

**Review by david-berthelot — CHANGES_REQUESTED** [review]
The code is almost there. 

Remember to add unit tests and documentation: https://objax.readthedocs.io/en/latest/dev/adding_module.html

Also you need to sign the Google CLA, otherwise we won't be allowed to accept your code.

Thanks!

File: objax/util/util.py (hunk 4 above)
Comment (naruto-raj) [reply objax/util/util.py:170]: sure @david-berthelot, i will look into them.

File: objax/constants.py (hunk 5 above)
Comment (naruto-raj) [reply objax/constants.py:26]: will add them too.

**Comment by naruto-raj** [comment]
> @googlebot I fixed it.

@googlebot I fixed it.

**Comment by naruto-raj** [comment]
@googlebot I fixed it.

**Comment by naruto-raj** [comment]
@googlebot I fixed it.

**Comment by naruto-raj** [comment]
@googlebot I consent.

**Comment by naveen-takvaviya** [comment]
@googlebot I consent.

**Comment by naruto-raj** [comment]
@david-berthelot  i will complete unit test and documentation next. Thank you for your guidance, Please let me know if any other change is required.

File: objax/functional/core/ops.py [hunk 6]
```diff
@@ -41,6 +44,27 @@ def flatten(x: JaxArray) -> JaxArray:
...
 
 
+def upsample_2d(x: JaxArray,
+                scale: Union[Tuple[int, int], int],
+                method: Union[UpSample, str] = UpSample.BILINEAR) -> JaxArray:
+    """Function to upscale 2D images.
+
+    Args:
+        x: input tensor.
+        scale: imt or tuple scaling factor
```
Comment (david-berthelot) [inline objax/functional/core/ops.py:54]:
```suggestion
        scale: int or tuple scaling factor
```

File: objax/util/util.py [hunk 7]
```diff
@@ -164,3 +165,13 @@ def to_tuple(v: Union[Tuple[Number, ...], Number, Iterable], n: int):
...
+
+
+def to_upsample(upsample: Union[UpSample, str]) -> Union[str]:
+    """Expand to a string method for interpolation"""
+    if isinstance(upsample, UpSample):
+        return upsample.value
+    if isinstance(upsample, str):
+        return UpSample[upsample.upper()].value
+
+    assert isinstance(upsample, (str, UpSample)), "Argument must be a string or UpSample"
```
Comment (david-berthelot) [inline objax/util/util.py:177]:
```suggestion
    assert isinstance(upsample, (str, UpSample)), f'Argument "{upsample}" must be a string or UpSample'
```

**Review by david-berthelot — CHANGES_REQUESTED** [review]
I just realized upsample_2d is a subcase of pure interpolation. It's perfectly fine to keep it since it makes life simpler but could you also add:
- an API `interpolate`, very similar in spirit to this one from PyTorch https://pytorch.org/docs/stable/nn.functional.html#interpolate

If I understand correctly, all it needs to do really is to wrap resize. If you'd rather not do it in this PR, we can create a new issue - I only propose this since I just realized given how similar in design it is. Let me know what you think.

**Comment by naruto-raj** [comment]
sure @david-berthelot ,i think its better to add interpolate as well. please create a new issue and i will start working on that asap.

**Comment by david-berthelot** [comment]
Added https://github.com/google/objax/issues/134

**Comment by shs037** [comment]
Hi @naruto-raj , is there any update on this pull request? Do you need any help?

**Comment by naruto-raj** [comment]
> Hi @naruto-raj , is there any update on this pull request? Do you need any help?

Hi @shs037  due to severe storms i didn't have a working internet connection for the past months, i would be getting back to developing the module and hopefully completing within the coming week and would be glad to have your help in completing it.

**Comment by shs037** [comment]
> > Hi @naruto-raj , is there any update on this pull request? Do you need any help?
> 
> Hi @shs037 due to severe storms i didn't have a working internet connection for the past months, i would be getting back to developing the module and hopefully completing within the coming week and would be glad to have your help in completing it.

Got it. Take your time and good luck with the internet! Let us know if we can help.

**Review by naruto-raj — COMMENTED** [review]
hi @david-berthelot and @shs037,  have a look at the commits and let me know if any change is required also the tests did not run on Github while they passed in my local. any idea what might be the reason?

**Review by david-berthelot — APPROVED** [review]
Thanks!

**Comment by david-berthelot** [comment]
Can you resolve the conflicts?

**Review by david-berthelot — APPROVED** [review]
Thanks!

**Review by david-berthelot — CHANGES_REQUESTED** [review]
Might be that you broke something while resolving conflicts. I see some tests are failing.
If it gets too confusing you can make a fresh branch and copy your changes to it.

File: objax/jaxboard.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (david-berthelot) [inline objax/jaxboard.py:1]: Seems you added an empty line on this file for some reason.

**Review by david-berthelot — CHANGES_REQUESTED** [review]

**Review by david-berthelot — APPROVED** [review]
Thanks!

File: objax/jaxboard.py (hunk 8 above)
Comment (naruto-raj) [reply objax/jaxboard.py]: sorry. didn't see that one.

**Comment by naruto-raj** [comment]
Thanks @david-berthelot, I should celebrate my first open-source contribution 🥇 .
